### PR TITLE
fix #568, [REST] add history metrics service in backend so that the UI d...

### LINKED ISSUE #568: [REST] provide history metrics service
The REST should allow user to query history metrics data, so that user can view the history metrics timecharts in UI dashboard.

### PR DESCRIPTION
**Pull request by clockfly**
...ashboard can query and present the timecharts which allow user to browser by specifying a time window.

[Review on Reviewable](https://reviewable.io/reviews/intel-hadoop/gearpump/572)

### CODE REVIEW

File: streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/HistoryMetricsService.scala [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+    override def add(inputMetrics: MetricType): Unit = {
+      history.add(inputMetrics)
+      recent.add(inputMetrics)
+    }
+
+    override def read: List[HistoryMetricsItem] = {
+      history.read ++ recent.read
+    }
+  }
+}
```
Comment (kkasravi) [inline streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/HistoryMetricsService.scala]: Should we keep RETAIN_* values in a conf somewhere? Also do we want a REST/WS interface to retrieve/change these values?

File: streaming/src/test/scala/org/apache/gearpump/streaming/appmaster/HistoryMetricsServiceSpec.scala [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    //expire oldest data point, because we only keep @count records
+    store.add(Counter("count", 5))
+    store.add(Counter("count", 6))
+
+    val result = store.read
+    assert(result.size == count)
+
+    //the oldest value is expired
+    assert(result.head.value.asInstanceOf[Counter].value == 3L)
+
```
Comment (huafengw) [inline streaming/src/test/scala/org/apache/gearpump/streaming/appmaster/HistoryMetricsServiceSpec.scala]: HistoryMetricsServiceSpec failed.

File: streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/HistoryMetricsService.scala (hunk 1 above)
Comment (clockfly) [inline streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/HistoryMetricsService.scala]: Currently, we store everything in memory. It is not safe at current point to expose this config to user. 
So, I prefer to make it static if for now.

Comment (clockfly) [inline streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/HistoryMetricsService.scala]: After rethinking this, I will make them configurable, and add a comment that this value should not be set to too big, due to implementation limit.

**Comment by clockfly** [comment]
@kkasravi 

comments fixed

**Comment by kkasravi** [comment]
Failed test, otherwise +1 !

**Pull request merged**